awslabs/soci-snapshotter
# Improve zTOC build errors

Signed-off-by: Kern Walster <walster@amazon.com>

*Issue #, if available:*

N/A

*Description of changes:*

Before this change, when trying to build a zTOC for an uncompressed
docker image layer, `soci` returned an unhelpful error:

```
soci: could not get index: -3
```

This is returning the raw error from `inflate` which fails because
the file is not compressed.

This change improves the error messages when zTOC creation fails by:
1. Checking layer compression type before attempting to build a zTOC
2. Including the layer digest in error messages (since zTOC builds are run in multiple
 goroutines, it is not always clear)
3. Adding additional context around gzip errors


*Testing performed:*

```
make check && make test && make integration
```

Manual tests:
Using this dockerfile:
```
FROM alpine:latest

RUN touch /var/soci
```

```
$ docker build -t example .
$ docker image save -o example.tar example
$ sudo ctr i import ./example.tar
$ sudo soci create docker.io/library/example:latest
soci: could not build zTOC for sha256:f5e8680d2a16fbf0f463b28376c15d72d2478b5003f35a8d60b7d7f7ac022f09: only layers compressed with gzip are supported
```

---

For anyone who comes across this as an actual error, the way to avoid it is to comrpess blobs when importing:
```
$ sudo ctr i import --compress-blobs ./example.tar
```

If you've already imported without compression, you will need to purge both the image and the content before importing again. E.g.
```
$ sudo ctr i rm docker.io/library/example:latest
$ sudo ctr content rm sha256:f5e8680d2a16fbf0f463b28376c15d72d2478b5003f35a8d60b7d7f7ac022f09
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
